Use `Eq1` in context heads.

When building with GHC-8.1.2017xxxx

```
[14 of 19] Compiling Control.Monad.Trans.Free.Church ( src/Control/Monad/Trans/Free/Church.hs, /home/ogre/Documents/kmettverse/dist-newstyle/build/x86_64-linux/ghc-8.1.20170123/free-5/build/Control/Monad/Trans/Free/Church.o )

src/Control/Monad/Trans/Free/Church.hs:83:10: warning: [-Wsimplifiable-class-constraints]
    The constraint ‘Eq (FreeT f m a)’ matches an instance declaration
    instance [safe] (Data.Functor.Classes.Eq1 f,
                     Data.Functor.Classes.Eq1 m, Eq a) =>
                    Eq (FreeT f m a)
```

We should rewrite those instances too, to use lifted classes, shouldn't we?

--- a/src/Control/Alternative/Free.hs
+++ b/src/Control/Alternative/Free.hs
@@ -6,6 +6,8 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 #endif
 {-# OPTIONS_GHC -Wall #-}
+#include "free-common.h"
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Alternative.Free

--- a/src/Control/Alternative/Free/Final.hs
+++ b/src/Control/Alternative/Free/Final.hs
@@ -1,4 +1,5 @@
 {-# LANGUAGE RankNTypes #-}
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Alternative.Free.Final

--- a/src/Control/Applicative/Free.hs
+++ b/src/Control/Applicative/Free.hs
@@ -5,10 +5,8 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 #endif
 {-# OPTIONS_GHC -Wall #-}
+#include "free-common.h"
 
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Applicative.Free

--- a/src/Control/Applicative/Free/Final.hs
+++ b/src/Control/Applicative/Free/Final.hs
@@ -1,9 +1,7 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE RankNTypes #-}
+#include "free-common.h"
 
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Applicative.Free.Final

--- a/src/Control/Applicative/Trans/Free.hs
+++ b/src/Control/Applicative/Trans/Free.hs
@@ -5,6 +5,8 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 #endif
 {-# OPTIONS_GHC -Wall #-}
+#include "free-common.h"
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Applicative.Trans.Free

--- a/src/Control/Comonad/Cofree/Class.hs
+++ b/src/Control/Comonad/Cofree/Class.hs
@@ -3,6 +3,7 @@
 {-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE UndecidableInstances #-}
+#include "free-common.h"
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Comonad.Cofree.Class

--- a/src/Control/Monad/Free/Church.hs
+++ b/src/Control/Monad/Free/Church.hs
@@ -4,9 +4,8 @@
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
+#include "free-common.h"
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Monad.Free.Church

--- a/src/Control/Monad/Free/Class.hs
+++ b/src/Control/Monad/Free/Class.hs
@@ -8,10 +8,8 @@
 {-# LANGUAGE TypeFamilies #-}
 #endif
 {-# OPTIONS_GHC -fno-warn-deprecations #-}
+#include "free-common.h"
 
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Control.Monad.Free.Class

--- a/src/Control/Monad/Free/TH.hs
+++ b/src/Control/Monad/Free/TH.hs
@@ -1,8 +1,5 @@
 {-# LANGUAGE CPP #-}
-
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
+#include "free-common.h"
 
 -----------------------------------------------------------------------------
 -- |

--- a/src/Control/Monad/Trans/Free/Church.hs
+++ b/src/Control/Monad/Trans/Free/Church.hs
@@ -3,14 +3,7 @@
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE UndecidableInstances #-}
-
-#ifndef MIN_VERSION_base
-#define MIN_VERSION_base(x,y,z) 1
-#endif
-
-#ifndef MIN_VERSION_mtl
-#define MIN_VERSION_mtl(x,y,z) 1
-#endif
+#include "free-common.h"
 
 -----------------------------------------------------------------------------
 -- |
@@ -70,7 +63,7 @@ import qualified Control.Monad.Trans.Free as FreeT
 import qualified Data.Foldable as F
 import qualified Data.Traversable as T
 import Data.Functor.Bind hiding (join)
-import Data.Function
+import Data.Functor.Classes.Compat
 
 #if !(MIN_VERSION_base(4,8,0))
 import Data.Foldable (Foldable)
@@ -80,11 +73,25 @@ import Data.Traversable (Traversable)
 -- | The \"free monad transformer\" for a functor @f@
 newtype FT f m a = FT { runFT :: forall r. (a -> m r) -> (forall x. (x -> m r) -> f x -> m r) -> m r }
 
-instance (Functor f, Monad m, Eq (FreeT f m a)) => Eq (FT f m a) where
-  (==) = (==) `on` fromFT
+#ifdef LIFTED_FUNCTOR_CLASSES
+instance (Functor f, Monad m, Eq1 f, Eq1 m) => Eq1 (FT f m) where
+  liftEq eq x y = liftEq eq (fromFT x) (fromFT y)
 
-instance (Functor f, Monad m, Ord (FreeT f m a)) => Ord (FT f m a) where
-  compare = compare `on` fromFT
+instance (Functor f, Monad m, Ord1 f, Ord1 m) => Ord1 (FT f m) where
+  liftCompare cmp x y= liftCompare cmp (fromFT x) (fromFT y)
+#else
+instance (Functor f, Monad m, Eq1 f, Eq1 m) => Eq1 (FT f m) where
+  eq1 x y = eq1 (fromFT x) (fromFT y)
+
+instance (Functor f, Monad m, Ord1 f, Ord1 m) => Ord1 (FT f m) where
+  compare1 x y = compare1 (fromFT x) (fromFT y)
+#endif
+
+instance (Eq1 (FT f m), Eq a) => Eq (FT f m a) where
+  (==) = eq1
+
+instance (Ord1 (FT f m), Ord a) => Ord (FT f m a) where
+  compare = compare1
 
 instance Functor (FT f m) where
   fmap f (FT k) = FT $ \a fr -> k (a . f) fr

